File: app/src/main/java/net/chris/demo/guardian/model/inject/NetworkModule.java
package net.chris.demo.guardian.model.inject;

import android.support.annotation.NonNull;

import com.fasterxml.jackson.databind.ObjectMapper;

import net.chris.demo.guardian.network.NetworkService;
import net.chris.lib.network.retrofit.BaseNetworkModuleHelper;

import javax.inject.Singleton;

import dagger.Module;
import dagger.Provides;
import okhttp3.HttpUrl;

@Module
public class NetworkModule {

    private BaseNetworkModuleHelper helper;

    public NetworkModule(@NonNull final HttpUrl baseUrl,
                         @NonNull final boolean debug) {
        helper = new NetworkModuleHelper(baseUrl, debug);
    }

    @Provides
    @Singleton
    protected NetworkService provideNetworkRequester() {
        return helper.getRetrofitBuilder().build().create(NetworkService.class);
    }

    @Provides
    @Singleton
    protected ObjectMapper provideObjectMapper() {
        return helper.getObjectMapper();
    }

}

File: app/src/main/kotlin/net/chris/demo/guardian/model/inject/NetworkModule.kt
package net.chris.demo.guardian.model.inject

import com.fasterxml.jackson.databind.ObjectMapper
import dagger.Module
import dagger.Provides
import net.chris.demo.guardian.network.NetworkService
import net.chris.lib.network.retrofit.BaseNetworkModuleHelper
import okhttp3.HttpUrl
import javax.inject.Singleton

@Module
class NetworkModule(baseUrl: HttpUrl,
                    debug: Boolean) {

    private var baseUrl = baseUrl

    private var debug = debug

    @Provides
    @Singleton
    fun provideNetworkRequester(helper: BaseNetworkModuleHelper): NetworkService = helper.retrofitBuilder.build().create(NetworkService::class.java)

    @Provides
    @Singleton
    fun provideObjectMapper(helper: BaseNetworkModuleHelper): ObjectMapper = helper.objectMapper

    @Provides
    @Singleton
    fun provideNetworkModuleHelper(): BaseNetworkModuleHelper = NetworkModuleHelper(baseUrl, debug)

}
